crates/env/src/engine/mod.rs

```diff
@@ -37,3 +41,109 @@ cfg_if! {
         }
     }
 }
+
+// The `Result` type used to represent the programmer defined contract output.
+type ContractResult<T, E> = core::result::Result<T, E>;
+
+// We only use this function when 1) compiling to Wasm 2) compiling for tests.
+#[cfg_attr(all(feature = "std", not(test)), allow(dead_code))]
+pub(crate) fn decode_fallible_constructor_reverted_return_value<I, E, ContractError>(
+    out_return_value: &mut I,
+) -> EnvResult<ConstructorResult<ContractResult<E::AccountId, ContractError>>>
+where
+    I: scale::Input,
+    E: crate::Environment,
+    ContractError: scale::Decode,
+{
+    let out: ConstructorResult<Result<(), ContractError>> =
+        scale::Decode::decode(out_return_value)?;
+
+    match out {
+        ConstructorResult::Ok(ContractResult::Ok(())) => {
+            // Since the contract reverted we don't expect an `Ok` return value from the
+            // constructor, otherwise we'd be in the `AccountId` decoding branch.
+            Err(crate::Error::Decode(
+                "TODO: probably shouldn't be a `Decode` error".into(),
```

Comment:
@ascjones I'm tempted to just panic here like we do in the infallible constructor case. It doesn't seem appropriate to let the caller handle this since it's either bad encoding by a) the codegen or b) the callee

Reply:
Panicking would be okay, the alternative would be to raise a `LangError`. It means we capture the reason why this failed, and is in keeping with the semantics of the top level `try_instantiate` method,

The question is whether it fits in the category of `LangError`s that we had in mind. I'm not 100% sure on this but there is at least a case to be made for it - it is a possible scenario which we can handle gracefully in this branch.